### ISSUE TITLE
DNS reverse lookup of embedded DNS server is broken

### ISSUE THREAD

**Issue by floragunncom** [docker/docker#26298]
<pre>
Client:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 21:15:28 2016
 OS/Arch:      darwin/amd64

Server:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 21:15:28 2016
 OS/Arch:      linux/amd64

Containers: 79
 Running: 2
 Paused: 0
 Stopped: 77
Images: 426
Server Version: 1.12.0
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 434
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null overlay bridge host
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.4.15-moby
Operating System: Alpine Linux v3.4
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 1.954 GiB
Name: moby
ID: ZN4U:JB7V:LNAY:5VCD:UMJU:DXG2:2RGC:5WHB:5JM3:PNGN:ZETO:C47V
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 41
 Goroutines: 99
 System Time: 2016-09-04T01:21:33.576370105Z
 EventsListeners: 2
No Proxy: *.local, 169.254/16
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
</pre>

Reverse DNS lookup performed by embedded DNS server (127.0.11) reports wrong and invalid hostname (like docker_hw2.example.com_1.docker_default).

Steps to reproduce the issue:

1) Create a docker-composer file
<pre>
version: '2'
services:
  hw1.example.com:
    image: httpd
    hostname: hw1.example.com
  hw2.example.com:
    image: httpd
    hostname: hw2.example.com
</pre>

2) Run ``docker-compose up``

3) Run ``docker exec docker_hw2.example.com_1 cat /etc/hosts`` and memorize ip address for hw2.example.com

4) Run ``docker exec docker_hw1.example.com_1 apt-get update``

5) Run ``docker exec docker_hw1.example.com_1 apt-get -yqq install host``

6) Run ``docker exec docker_hw1.example.com_1 host -v <IP ADDRESS>``

Result:
<pre>
Trying "3.0.18.172.in-addr.arpa"
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 27786
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;3.0.18.172.in-addr.arpa.      	IN     	PTR

;; ANSWER SECTION:
3.0.18.172.in-addr.arpa. 600   	IN     	PTR    	docker_hw2.example.com_1.docker_default.

Received 117 bytes from 127.0.0.11#53 in 1 ms
</pre>

``docker_hw2.example.com_1.docker_default`` is a wrong and invalid hostname

Its wrong because it should be 'hw2.example.com' and invalid because hostname must not contain underline (_) character according to RFC 1123, section 2.1

In my case this breaks SSL hostname verification.

**Comment by vieux** [comment]
@mrjana can you please take a look ?

**Comment by sanimej** [comment]
@floragunncom Service discovery in Docker has always been network scoped; ie: a container attached to a network can discover other containers in the same network. There is no global namespace for container names. The FQDN for a service name is `container_name.network_name`. The behavior you are seeing is expected. The reason you see `_` in the name is because of the way compose names the containers.

Do you have the ssl client & server running in containers ?

**Comment by floragunncom** [comment]
Yes, in one container a ssl client is running and in another container a ssl server is running. I try to establish a ssl connection with hostname resolution and verification.

**Comment by ChristianKniep** [comment]
+1 on this issue. 
I am experimenting with docker service to spin up HDFS and the hostname generated by doker service is comprised of `${SERVICE_NAME}.${SLOT_ID}.${TASK_ID}.${NETWORK_NAME}` flips hadoop over.

```
java.lang.IllegalArgumentException: Does not contain a valid host:port authority: hdfs-namenode.1.1ub2hholjekbir0396m87iby7:8020
```
This seems to stem from the RFC, which rejects the toplabel of the FQDN: http://www.ietf.org/rfc/rfc2396.txt

```
hostport      = host [ ":" port ]
host          = hostname | IPv4address
hostname      = *( domainlabel "." ) toplabel [ "." ]
domainlabel   = alphanum | alphanum *( alphanum | "-" ) alphanum
toplabel      = alpha | alpha *( alphanum | "-" ) alphanum
```

So this conflicts with the standard as the `TASK_ID` is not starting (necessarily) with an alpha char.

**Comment by ChristianKniep** [comment]
Any news on this front?

**Comment by sanimej** [comment]
@floragunncom As [clarified earlier](https://github.com/docker/docker/issues/26298#issuecomment-245026447) docker service discovery is based on the container names, not host names. And containers belong to the domain identified by the network name. 

Compose v2 creates a default network `docker_default` if you don't specify a network name. But you can explicitly name the network to be created under `networks:` You can also specify the container name using `container_name:`

In your case since you want the PTR query to return the hostname you can achieve it by setting the container name, network name and host name appropriately. For ex: `container_name: web`, `network: prod` and `hostname: web.prod`.

Other option is to run an external DNS server that resolves based on the docker's hostname. Though this needs the PR https://github.com/docker/libnetwork/pull/1586 to be merged and available in docker.

**Comment by sanimej** [comment]
@ChristianKniep `cbf1d1439814` in the hosts file is the shortened ID of the container. Thats because In 1.12 hostname of the container is its short ID. In 1.13 you can pass `--hostname` option to `service create`. All the tasks of that service will get that hostname. So the reverse lookup for container's IP will give the user provided hostname. That might help you.

PTR query for some other container's IP will still return `container_name.network_name`.  That is the naming scheme used for Docker service discovery. [As I mentioned earlier](https://github.com/docker/docker/issues/26298#issuecomment-265620489), one option to achieve what you want is to run an external DNS server that provides the resolution and reverse lookup using the hostname.

**Comment by dsevilla** [comment]
I am also locked with this issue, as I don't have any way of fixing it apart from using an external DNS, which is infortunate. For some reason some reverse DNS requests are picking the name augmented with the network name, which makes, for example, Zookeeper not working (and the whole Hadoop/HBase stack). Even providing a hostname and domainname, the network name is added, which IMHO it's a bug.

**Comment by OOPMan** [comment]
As of now, reverse DNS requests don't work at all.

**Comment by pan3793** [comment]
I have the same issue in docker compose v2, any workaround in 2024?
